### PR TITLE
Add `test` arg to AsyncValue.guard

### LINKED ISSUE #3291: Support to specify exception type on `AsyncValue.guard`
**Is your feature request related to a problem? Please describe.**

We are facing some unexpected situations when using `AsyncValue.guard`. We do not want to guard exceptions that we would like the application to crash, like a null pointer exception. For these exceptions we want that the application crash and the error gets propagated to the upper layers, like our crash tool layer.

**Describe the solution you'd like**
We would like an way to specify what type of exception to guard, like:

```dart
static Future<AsyncValue<T>> guard<T, E extends Object>(Future<T> Function() future) async {
    try {
      return AsyncValue.data(await future());
    } on E catch (err, stack) {
      return AsyncValue.error(err, stack);
    }
  }
```

**Describe alternatives you've considered**
We can't use Dart extensions because we can't define static methods for extensions. What we are considering to do is to create a global function that does the job, something like:

```dart
Future<AsyncValue<T>> exceptionGuard<T, E extends Object>(Future<T> Function() future) async {
  try {
    return AsyncValue.data(await future());
  } on E catch (err, stack) {
    return AsyncValue.error(err, stack);
  }
}
```

and instead of doing `state = await AsyncValue.guard(future);` we do `state = await exceptionGuard(future);`

We also face a similar issue with the `build` method in `AsyncNotifier` classes. The function automatically capture exceptions that happen on its body execution, but what if we want that some of these exceptions is thrown?

Let me know your thoughts about this feature request.

### PR DESCRIPTION
**Pull request by utamori**
## Related Issues

fixes #3291

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).

- [x] I have updated the `CHANGELOG.md` of the relevant packages.
      Changelog files must be edited under the form:

  ```md
  ## Unreleased fix/major/minor

  - Description of your change. (thanks to @yourGithubId)
  ```

- [x] If this contains new features or behavior changes,
      I have updated the documentation to match those changes.

### CODE REVIEW

**Comment by rrousselGit** [comment]
Thanks for this!
Make sure to update the changelog and docs, as per the issue template :)

File: packages/riverpod/lib/src/common.dart [hunk 1]
```diff
@@ -157,12 +157,24 @@ abstract class AsyncValue<T> {
   ///     });
   ///   }
   /// }
+  ///
+  /// when predicate is provided, and it is false; rethrown.
```
Comment (rrousselGit) [inline packages/riverpod/lib/src/common.dart:161]: You've edited the code-snippet. Explanation should be outside of the ` ``` `

Also, could you expand on it a bit more?

Comment (utamori) [reply packages/riverpod/lib/src/common.dart]: Thanks, I added it.

File: packages/riverpod/lib/src/common.dart [hunk 2]
```diff
@@ -157,12 +157,43 @@ abstract class AsyncValue<T> {
   ///     });
   ///   }
   /// }
+  ///
+  /// By using the [test] argument, we can add additional conditions.
```
Comment (rrousselGit) [inline packages/riverpod/lib/src/common.dart:161]:
```suggestion
  /// An optional callback can be specified to catch errors only under a certain condition.
  /// In the following example, we are only catching FormatExceptions.
```

File: packages/riverpod/lib/src/common.dart [hunk 3]
```diff
@@ -157,12 +157,33 @@ abstract class AsyncValue<T> {
   ///     });
   ///   }
   /// }
+  ///
+  /// An optional callback can be specified to catch errors only under a certain condition.
+  /// In the following example, we catch all exceptions beside FormatExceptions.
+  ///
+  /// ```dart
+  ///   AsyncValue.guard(
+  ///    () async { /* ... */ },  ///     },
```
Comment (rrousselGit) [inline packages/riverpod/lib/src/common.dart:166]:
```suggestion
  ///    () async { /* ... */ },
```

**Comment by rrousselGit** [comment]
LGTM, thanks!

**Pull request merged**